SPARK-1714. Take advantage of AMRMClient APIs to simplify logic in YarnA...

sryza opened the pull request:
...llocator

The goal of this PR is to simplify YarnAllocator as much as possible and get it up to the level of code quality we see in the rest of Spark.

In service of this, it does a few things:
- Uses AMRMClient APIs for matching containers to requests.
- Adds calls to AMRMClient.removeContainerRequest so that, when we use a container, we don't end up requesting it again.
- Removes YarnAllocator's host->rack cache. YARN's RackResolver already does this caching, so this is redundant.
- Adds tests for basic YarnAllocator functionality.
- Breaks up the allocateResources method, which was previously nearly 300 lines.
- A little bit of stylistic cleanup.
- Fixes a bug that causes three times the requests to be filed when preferred host locations are given.

The patch is lossy. In particular, it loses the logic for trying to avoid containers bunching up on nodes. As I understand it, the logic that's gone is:
- If, in a single response from the RM, we receive a set of containers on a node, and prefer some number of containers on that node greater than 0 but less than the number we received, give back the delta between what we preferred and what we received.

This seems like a weird way to avoid bunching E.g. it does nothing to avoid bunching when we don't request containers on particular nodes.

sryza committed:
SPARK-1714. Take advantage of AMRMClient APIs to simplify logic in YarnAllocator

sryza committed:
Further documentation and cleanup

sryza committed:
Take out rack stuff

sryza committed:
Remove locality stuff and more cleanup

sryza committed:
Fix YarnClusterSuite

sryza committed:
Simplify further

sryza committed:
Fix scalastyle

sryza committed:
Fix a couple comments and simplify requestTotalExecutors

sryza commented:
@tgravescs are you able to take a look at this?

lianhuiwang commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -115,26 +107,35 @@ private[yarn] class YarnAllocator(
-      logInfo(s"Not allocating more executors because there are already $currentTotal " +
-        s"(application requested $requestedTotal total)")
-    }
+    maxExecutors = requestedTotal
```
i think there maybe is maxExecutors += requestedTotal. because maxExecutors is sum of current executors, including running and pending executors.

lianhuiwang commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -153,498 +154,241 @@ private[yarn] class YarnAllocator(
-          }
+        // Decrement the number of executors running. The next iteration of
+        // the ApplicationMaster's reporting thread will take care of allocating.
+        numExecutorsRunning -= 1
```
i think there maxExecutors  should -= 1

sryza replied:
`maxExecutors` is the total number of executors that we want to be running.  A container completion doesn't mean that we want fewer executors - we'd like to request another executor to replace the fallen one.

lianhuiwang replied:
so i think we should rename maxExecutors to maxRequestExecutors that can be better to represent its meaning.

lianhuiwang commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -62,50 +62,42 @@ private[yarn] class YarnAllocator(
     amClient: AMRMClient[ContainerRequest],
     appAttemptId: ApplicationAttemptId,
     args: ApplicationMasterArguments,
-    preferredNodes: collection.Map[String, collection.Set[SplitInfo]],
```
if we remove preferredNodes, so locality cannot be work. because container maybe schedule any nodemanager.

sryza replied:
This is discussed in the proposal - preferred locality on YARN is currently broken for a few other reasons.  The goal here is make the code as simple as possible, and we can later add it back in in a way that makes sense for SPARK-4352.

lianhuiwang commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -153,498 +154,241 @@ private[yarn] class YarnAllocator(
   def allocateResources(): Unit = synchronized {
-    val missing = maxExecutors - numPendingAllocate.get() - numExecutorsRunning.get()
+    val numPendingAllocate = getNumPendingAllocate
+    val missing = maxExecutors - numPendingAllocate - numExecutorsRunning
```
if maxExecutors is the total number of executors that we want to be running, so there missing = maxExecutors - numPendingAllocate?

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -115,26 +107,35 @@ private[yarn] class YarnAllocator(
 
   /**
    * Request as many executors from the ResourceManager as needed to reach the desired total.
    * This takes into account executors already running or pending.
```
perhaps update this comment since we are no longer calculating based on the num pending, executing, etc.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -153,498 +154,241 @@ private[yarn] class YarnAllocator(
-      val totalExecutorMemory = executorMemory + memoryOverhead
-      numPendingAllocate.addAndGet(missing)
-      logInfo(s"Will allocate $missing executor containers, each with $totalExecutorMemory MB " +
+      logInfo(s"Will request $missing executor containers, each with ${resource.getMemory} MB " +
```
it might be nice to also print the cores here.   I know you didn't change this but if you happen to be changing stuff there

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -153,498 +154,241 @@ private[yarn] class YarnAllocator(
-  private def internalReleaseContainer(container: Container): Unit = {
-    releasedContainers.put(container.getId(), true)
-    amClient.releaseAssignedContainer(container.getId())
+    logInfo("Received %d containers from YARN, launching executors on %d."
```
changing launching to launched

sryza replied:
Because of the async nature of `runAllocatedContainers`, no communication with YARN has necessarily occurred, so I think "launched" would be inaccurate.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -153,498 +154,241 @@ private[yarn] class YarnAllocator(
-   * @param count Number of containers to allocate.
-   *              If zero, should still contact RM (as a heartbeat).
-   * @return Response to the allocation request.
+   * @param containersToUse list of containers that will be used
```
add description for the other 2 params

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -153,498 +154,241 @@ private[yarn] class YarnAllocator(
-    for (request <- containerRequests) {
-      amClient.addContainerRequest(request)
-    }
+      allocatedHostToContainersMap.synchronized {
```
is this synchronization really needed?  It looks like the only times these structures are updated are in runAllocated and processCompleted which are both called from synchronized function allocateResources

tgravescs replied:
nevermind, I assume its for the tests

sryza replied:
Looks like it's not

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala:
```diff
@@ -85,7 +85,7 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
 }
 
-object YarnSparkHadoopUtil {
+private[spark] object YarnSparkHadoopUtil {
```
I don't think we can change this to private at this point. Already public in 1.0

tgravescs commented:
@sryza  did you run this through a bunch of manual tests as well?  Try cases like container dies/killed before all of the first ones allocated.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -28,14 +29,13 @@ import com.google.common.util.concurrent.ThreadFactoryBuilder
+import org.apache.spark.scheduler.SplitInfo
 import org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend
 
 object AllocationType extends Enumeration {
```
this isn't used anymore - you can remove it

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -43,18 +43,18 @@ object AllocationType extends Enumeration {
+ *
+ * This class makes use of YARN's AMRMClient APIs. We interact with the AMRMClient in three ways:
+ * * Making our resource needs known, which updates local bookkeeping about containers requested.
+ * * Calling "allocate", which syncs our local container requests with the RM, and returns any
```
this is also a regular heartbeat

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -153,498 +154,241 @@ private[yarn] class YarnAllocator(
+   */
+  def addResourceRequests(numExecutors: Int): Unit = {
+    val containerRequests = new ArrayBuffer[ContainerRequest]
+    for (i <- 0 until numExecutors) {
```
any particular reason we are doing 3 loops here instead of just doing all of this in 1 loop?  create, addContainerRequest, print info

sryza replied:
Looks like not.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -153,498 +154,241 @@ private[yarn] class YarnAllocator(
-        val rackCount = rackToCount.getOrElse(host, 0)
-        rackToCount.put(host, rackCount + splits.size)
+    if (!remainingAfterOffRackMatches.isEmpty) {
+      for (container <- remainingAfterOffRackMatches) {
```
it might be nice to have debug statement here about releasing extra container

SparkQA commented:
[Test build #25850 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25850/consoleFull) for   PR 3765 at commit [`74f56dd`](https://github.com/apache/spark/commit/74f56dd6bcc8248de061bb1567a9c966b90d080d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25850 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25850/consoleFull) for   PR 3765 at commit [`74f56dd`](https://github.com/apache/spark/commit/74f56dd6bcc8248de061bb1567a9c966b90d080d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25850/
Test PASSed.

sryza committed:
Muffle RackResolver logs

sryza commented:
@tgravescs, uploaded a new patch that addresses your review comments.  I just ran a bunch of manual tests on a 6-node, including
* request more resources than the cluster has available
* kill executors while a job is running
* kill executors before we've received the full set of initial executors

I noticed that YARN's `RackResolver`, which we now rely on directly, is really noisy, so I added a line in log4j-defaults.properties to muffle those logs.

SparkQA commented:
[Test build #25857 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25857/consoleFull) for   PR 3765 at commit [`32a5942`](https://github.com/apache/spark/commit/32a594295e97f74bda48dfb026bbb842163607ae).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25857 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25857/consoleFull) for   PR 3765 at commit [`32a5942`](https://github.com/apache/spark/commit/32a594295e97f74bda48dfb026bbb842163607ae).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25857/
Test PASSed.

tgravescs commented:
looks good. Thanks @sryza
